### ISSUE TITLE
filesystem check loop broken?

### ISSUE THREAD

**Issue by borkd** [lizardfs/lizardfs#298]
Filesystem check loop seems to be subtly broken. Information presented in "info" and "chunks" tabs has not been consistent for the last couple of days. A number of under-goal files and chunks was reported in "info tab", yet not a single filename shows up in the info tab. At the same time "chunks" tab reported all chunks as replicated per assigned goal levels. I have posted screenshots and more details in #220 and #222.

Please consider wrapping the code used for the filesystem check loop in a command line utility. This could help users troubleshoot issues or force scans as needed, but would also enable more test cases during development cycle.

**Comment by borkd** [comment]
After removing a snapshot with 5M+ files overnight I could see a bunch of endangered chunks and one chunk reported as missing in both info and chunks tab. Replication goals for affected files were set at 5 and we had no server outages. Filesystem check loop has not run since then so the name of the file with missing chunk has not shown up yet in the info page. 

I had to resort to running mfscheckfile on find results with 0 copies. First file found happened to be a fsprobe test target named fsprobe.

# mfscheckfile fsprobe
fsprobe:
 chunks with 0 copies:            1
 chunks with 3 copies:            1

# mfsfileinfo fsprobe
fsprobe:
        chunk 0: 00000000011F3DEA_00000002 / (id:18824682 ver:2)
                copy 1: ip1:9422:label1
                copy 2: ip2:9622:label2
                copy 3: ip3:9422:label3

# mfsgetgoal fsprobe
fsprobe: x3

# grep x3 /etc/mfs/mfsgoals.cfg
x3 : xfc _ _ _ _

# hexdump -C fsprobe
00000000  f0 f0 f0 f0 f0 f0 f0 f0  f0 f0 f0 f0 f0 f0 f0 f0  |................|
*
hexdump: fsprobe: No such device or address
04000000

As I reported earlier (see my comments in #222 and #220) filesystem check loop indicated for a while that some files and chunks were undergoal in the summary field of the info tab. To the contrary the chunk matrix and info on the chunks tab would indicate all is good. Mfsmaster runs on solid, fully redundant server hardware including memory mirroring, and the hardware reported no issues in the three months or so I have tested LizardFS. 

The fact that this reoccurring breakage occurs at a fraction of scale what normal usage would be my confidence in viability of LizardFS went from pretty strong to an all time low. Lack of response from core developers doesn't help either :(

@marcinsulikowski - any suggestions?

**Comment by psarna** [comment]
Hi borkd,

our *core developers lack of response* was due to very busy period, we will study your issue carefully next week - after we deal with other urgent things.

Meanwhile you can take a look at this upcoming patch: <http://cr.skytechnology.pl:8081/#/c/1963/>, which should solve some issues with chunks that have different goals from several snapshots.

Did your problem occur on chunks that were shared between snapshots that had different goals?

**Comment by borkd** [comment]
@psarna - some of the files most likely yes, fsprobe file mentioned above was bumped from 3 copies to 5 copies before getting lost.

**Comment by borkd** [comment]
@psarna - two questions:

1. besides not removing snapshots what is the recommendation to minimize chunk loss until http://cr.skytechnology.pl:8081/#/c/1963/ is merged into stable branch? My filesystem is only 49% full, yet a chunk with 5-way replication goal went missing.

2. After adding another chunkserver to one of the labels millions of chunks are apparently need of replication (to rebalance I guess). Any suggestions as to how to minimize "disconnected" messages reported in #220 that abound in master's logs? What are the odds that this replication will end up loosing chunks in the process?

Appreciate your efforts, thanks!

**Comment by psarna** [comment]
Well, first of all, you should verify if everything is fine with connection between chunkservers - "disconnected" messages suggest that something prevents replication from succeeding.  We will look closer at this issue in a couple of days.

As for minimizing chunk loss, I think that you should restraing from doing anything that causes mass replication, since something is clearly wrong with your installation (judging from lots of "disconnected" messages).

**Comment by DarkHaze** [comment]
Could you paste a list of your goals ( if that is not confidential ) in here ?
It would help us greatly debug the problem.

**Comment by borkd** [comment]
@DarkHaze - current version of the mfsgoals.cfg file may not be enough to reproduce the issue as the list of goals, and goal definitions themselves were evolving over time, and the system was upgraded from 2.5.4 to 2.6.0. I will try my best to document the configuration and help resolve this and other issues. See below for configuration overview:

1. mfsexports.cfg allows clients to mount a subdirectory /subdir of the exported filesystem as opposed to the root of the filesystem
2. mfstopology.cfg has four server groups configured in it and I have 7 custom chunkserver labels in use at this point.
3. Both default and custom goals were used for subdirectories of client visible directory tree. Some example goals:
```
1 1 : raid
2 2 : scratch scratch
[...]
13 x3 : dc1 dc2 dc3
14 x5 : raid _ _ _ _
15 xall : raid _ _ _ _ _
# temporary goals for fast import of data
16 bulk : in in
17 bulkplus: in in _
[...]
```
4. mfsmakesnapshot was used periodically on mfsmaster to snapshot the contents of /subdir into /snapshot/TIMESTAMP directory, outside of reach of the clients. I do not recall changing goals for /snapshot/ directory or its contents. It takes over an hour to remove a single snapshot, and the mfsmaster jumps to 60-80% cpu load during that time. Snapshot removal was therefore done rarely, by hand, at opportune times. 

Note 1 - bulk data import: Clients running on the same systems as the chunkservers with label "in" bring in millions of files worth of new data. Once imported, the data will have their final goals set to let lizardfs rebalance as needed.

Note 2 - definition of goal x3 was changed to work around problems with one of the failure domains, then once again following "invalid copies" incident I reporded on Jun 19th as part of #222. x3 was first changed to
13 x3 : xfc dc1 dc2 dc3
then eventually to 
13 x3 : xfc _ _ _ _

Hope this helps a bit.

**Comment by borkd** [comment]
Why so many different goals: when dealing with more complex network topology and storage setup one needs work around current limitations of mfstopology.cfg (linear model, hard to define branches) and mfsgoals.cfg (any chunkserver or explicit label, but no wildcards that could encompass part of the topology in them). #269 suggests this could improve in the future.

**Comment by psarna** [comment]
OK, there could be several reasons for this particular failure:

1. Until http://cr.skytechnology.pl:8081/#/c/1963/ gets merged, handling multiple goals for chunks is very unintuitive, because only goal with **highest id** is effective.
An illustrative example:
    * File *file1* has goal *x3* (xfc _ _ _ _)
    * Snapshot is created: mfsmakesnapshot *file1* *file2*
    * Goal *bulk* (in in) is assigned to *file2*
    * Now chunks from *file1* have effective goal *bulk*, because they are shared with *file2* and *bulk* happens to have **higher id** than *x3*, which has nothing to do with their safety level.

    So, because of counter-intuitive goal handling some data could actually have lower goal than expected.
You can try to verify if it might have caused problems in CGI, by looking at Chunks tab,
where you can find effective goals of your chunks.
 
2. I assume that some of your chunkservers are faulty, judging from flood of "disconnected" replication statuses. Do they come with other error messages? Studying master & chunkserver logs could help recognize the issue.

Malfunctioning chunkservers(2) combined with understated goals(1) could possibly lead to undergoal chunks.

And for your previous question about "What are the odds that this replication will end up losing chunks in the process?" - the only odds I see is that:
   * chunkserver receives replicated data
   * chunkserver confirms receiving data
   * chunkserver loses data before flushing it to disk due to some failure
   * data is removed from the other chunkserver, because the one that crushed confirmed receiving data
This scenario is drastically improbable, unless mentioned chunks have low goal and some chunkservers are faulty (which might have been the case...).

**Comment by borkd** [comment]
@psarna: that explains a lot. I was about to open a new bug, but it all seems related now:

While trying to reshuffle data via goal change it appears that goal id is given more consideration that should be the case . System failed to switch goals when snapshots were present. To reproduce: 
 1. create/pick a custom goal with high id (bulk)
 2. create subdir with high goal id, set goal to bulk and populate with data
 3. mfssnapshot subdir testsnap (succeeds)
 4. check chunks tab for number of chunks with goal bulk (number of bulk chunks matches amount of data)
 5. create/pick a new goal with id lower than original (failtest)
 6. mfssetgoal -r failtest subdir (succeeds)
 7. mfsgetgoal -r subdir (shows failtest)
 8. check chunks tab for number of chunks with goal bulk and failtes (failtest will not even show up)
 9. date > subdir/foo
 10. check chunks tab for number of chunks with goal bulk and failtes (failtest shows 1 chunk from newly created file, chunks with goals changed via command line do not show up)

If you skip step number 3 all works well. 

TL;DR: If the documentation mentioned that numeric **goal id**  is given more consideration in terms of importance than corresponding **goal definition** I would do my best not to shoot myself in the foot :)

**Comment by psarna** [comment]
Proper documentation will be published soon, and a patch covering this goal id mess even sooner ;)
